### PR TITLE
Support more optionals

### PR DESCRIPTION
**Pull request by jbescos**
Adding support for OptionalInt, OptionalDouble and OptionalLong.

This was mentioned by @joschi here: https://github.com/eclipse-ee4j/jersey/issues/4651

### CODE REVIEW

**Review by senivam — APPROVED** [review]
this does make sense, but we have to deal with another PR on the same topic which is #4799 thus we will somehow merge those 2 PRs after fixing conflicts which would appear after one or another PR is merged. So, wait with merging but legitimate.

**Comment by jbescos** [comment]
@joschi I just merged your changes with mine. Would you mind to review it please?.

File: core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java [hunk 1]
```diff
@@ -305,6 +309,125 @@ public String toString(T value) throws IllegalArgumentException {
...
+
+            OPTIONAL(Optional.class) {
+                @Override
+                Object empty() {
+                    return Optional.empty();
+                }
+                @Override
+                Object of(Object value) {
+                    return Optional.of(value);
+                }
```
Comment (joschi) [inline core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java:380]: Do we need to cover this case? I think `Optional<T>` would already be handled by `OptionalCustomProvider` before `OptionalProvider` gets invoked.

File: tests/e2e-server/src/test/java/org/glassfish/jersey/tests/e2e/server/OptionalParamConverterNoProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+        @GET
+        @Path("/fromList")
+        public Response fromList(@QueryParam(PARAM_NAME) List<Optional<Integer>> data) {
+            StringBuilder builder = new StringBuilder("");
+            for (Optional<Integer> val : data) {
+                builder.append(val.orElse(0));
+            }
+            return Response.ok(builder.toString()).build();
+        }
```
Comment (joschi) [inline tests/e2e-server/src/test/java/org/glassfish/jersey/tests/e2e/server/OptionalParamConverterNoProviderTest.java:64]: Would it make sense to have a corresponding test for `List<OptionalInt>>`, `List<OptionalLong>>`, and `List<OptionalDouble>>`?

File: core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java [hunk 3]
```diff
@@ -305,6 +309,125 @@ public String toString(T value) throws IllegalArgumentException {
...
+                            T val = aggregated.getConverter((Class<T>) type, type, annotations).fromString(value.toString());
+                            if (val != null) {
+                                return (T) optionals.of(val);
+                            } else {
+                                /*
+                                 *  In this case we don't send Optional.empty() because 'value' is not null.
+                                 *  But we return null because the provider didn't find how to parse it.
+                                 */
+                                return null;
+                            }
```
Comment (joschi) [inline core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java:350]: If `Optionals` covered only `OptionalInt`, `OptionalLong`, and `OptionalDouble` we wouldn't need to cover this case (which is already covered in `CustomOptionalProvider`).

File: core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java (hunk 1 above)
Comment (jbescos) [reply core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java]: Right, it is not necessary

File: tests/e2e-server/src/test/java/org/glassfish/jersey/tests/e2e/server/OptionalParamConverterNoProviderTest.java (hunk 2 above)
Comment (jbescos) [reply tests/e2e-server/src/test/java/org/glassfish/jersey/tests/e2e/server/OptionalParamConverterNoProviderTest.java:64]: Done

File: core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java (hunk 3 above)
Comment (jbescos) [reply core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java]: Right, it is not necessary

**Review by joschi — APPROVED** [review]
LGTM. 👍

**Review by jansupol — APPROVED** [review]

**Pull request merged**